### ISSUE TITLE
Monkeypatching for tests?

### ISSUE THREAD

**Issue by diamondburned** [Xuanwo/go-locale#3]
https://github.com/Xuanwo/go-locale/blob/master/locale_linux_test.go#L16

Why?

Just. Why?

**Comment by diamondburned** [comment]
P/S: The library's license:

```
Copyright Bouke van der Bijl

I do not give anyone permissions to use this tool for any purpose. Don't use it.

I’m not interested in changing this license. Please don’t ask.
```

**Comment by Xuanwo** [comment]
Our unittests intend to test our logic without environmental interference. 

As you can see, `detectViaEnv`'s result depends on multiple ENVs and outer command `locale`. So I monkey patch them for unittest.

And there is a plan to add BDD integration tests to make sure this lib works correctly.

**Comment by diamondburned** [comment]
You could use a function variable for this, I think.

**Comment by Xuanwo** [comment]
Can you give me an example for that?

**Comment by diamondburned** [comment]
PR inbound...

**Comment by diamondburned** [comment]
https://github.com/Xuanwo/go-locale/pull/4

**Comment by Xuanwo** [comment]
I spent some time to dig into the concerning on monkey patching for test, especially on this project.

There are following problems on it:

- Monkey patching changes function behavior, could be confusing after real function behavior changed.
- This project built upon golang, on which monkey patching is not stable and could be fail at any time.
- The lib we used for monkey patching doesn't have valid license.

Those problems are not previously considered, let answer one by one:

- The function monkey patched including following state
  -  `os.LookupEnv`...: Stable in go 1
  - platform related call, `locale`, `defaults`: Stable in os level
  - self maintained function, `detectViaEnv`: Will be updated as the same time.
- Yep, but it works for now, and I will switch to another implementation while it failed.
- Yep, I can fork it to have a valid license, `bou.ke/monkey` is not a huge lib and only be used in testing.

As presented in PR #4 , without monkey patch, we still need hook/unhook logic for testing.

Thank @diamondburned ~

**Comment by diamondburned** [comment]
Hook/unhook logic is only optional and is used for testing. Monkeypatching is very ugly in Go, as it utilizes a lot of unsafe code (literally unsafe code). Hook and unhooking is a sane behavior that's used in the stdlib.

**Comment by diamondburned** [comment]
As said in its README, clearly anything with these kinds of restrictions shouldn't ever be used seriously:

> Monkey sometimes fails to patch a function if inlining is enabled. Try running your tests with inlining disabled, for example: go test -gcflags=-l. The same command line argument can also be used for build.
> Monkey won't work on some security-oriented operating system that don't allow memory pages to be both write and execute at the same time. With the current approach there's not really a reliable fix for this.
> Monkey is not threadsafe. Or any kind of safe.


Also bonus:

![image](https://user-images.githubusercontent.com/8463786/74809331-ca920180-52a1-11ea-87b7-d10b5d516fbf.png)

**Comment by Xuanwo** [comment]
Althougth monkey patching looks good for me, but this issue encourages me to refactor the whole lib so that we can get much more clear code.

Welcome to review my PR: https://github.com/Xuanwo/go-locale/pull/5
And https://github.com/Xuanwo/go-locale/pull/4 will be closed after #5 get merged.

**Comment by diamondburned** [comment]
very unepic but ok
